src/uct/tcp/tcp_ep.c

```diff
@@ -329,132 +421,200 @@ void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+         * EP as failed */
         uct_tcp_ep_set_failed(ep);
-        goto out;
     }
+    goto out;
+}
+
+void uct_tcp_ep_replace_ep(uct_tcp_ep_t *to_ep, uct_tcp_ep_t *from_ep)
+{
+    uct_tcp_iface_t *iface   = ucs_derived_of(to_ep->super.super.iface,
+                                              uct_tcp_iface_t);
+    int events               = from_ep->events;
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+
+    uct_tcp_ep_mod_events(from_ep, 0, from_ep->events);
+    to_ep->fd   = from_ep->fd;
+    from_ep->fd = -1;
+    uct_tcp_ep_mod_events(to_ep, events, 0);
 
-    if (*ep_p == NULL) {
-        *ep_p = ep;
+    to_ep->conn_retries++;
+
+    uct_tcp_ep_ctx_move(&to_ep->tx, &from_ep->tx);
+    uct_tcp_ep_ctx_move(&to_ep->rx, &from_ep->rx);
+
+    ucs_queue_splice(&to_ep->pending_q, &from_ep->pending_q);
+    ucs_queue_splice(&to_ep->put_comp_q, &from_ep->put_comp_q);
+
+    to_ep->flags |= from_ep->flags & (UCT_TCP_EP_FLAG_ZCOPY_TX           |
+                                      UCT_TCP_EP_FLAG_PUT_RX             |
+                                      UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK |
+                                      UCT_TCP_EP_FLAG_PUT_RX_SENDING_ACK);
+
+    if (uct_tcp_ep_ctx_buf_need_progress(&to_ep->rx)) {
+        /* If some data was already read, we have to process it */
+        uct_worker_progress_register_safe(&iface->super.worker->super,
+                                          uct_tcp_ep_progress_data_rx, to_ep,
+                                          UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
     }
 
-out:
-    return status;
+    /* The internal EP is not needed anymore, start failed flow for the
+     * internal EP in order to destroy it from progress (to not dereference
+     * already destroyed EP) */
+    ucs_assert(!(from_ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX));
+    uct_tcp_ep_set_failed(from_ep);
 }
 
-static ucs_status_t uct_tcp_ep_create_connected(uct_tcp_iface_t *iface,
-                                                const struct sockaddr_in *dest_addr,
-                                                uct_tcp_cm_conn_sn_t conn_sn,
-                                                uct_tcp_ep_t **ep_p)
+static ucs_status_t uct_tcp_ep_connect(uct_tcp_ep_t *ep)
 {
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
     ucs_status_t status;
+    uct_tcp_ep_t *peer_ep;
+
+    ucs_assert(uct_tcp_ep_is_unconnected(ep));
+
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTING);
+    if (uct_tcp_ep_is_self(ep) ||
+        (ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP) ||
+        ((peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->cm_id.conn_sn,
+                                      UCT_TCP_EP_FLAG_CTX_TYPE_RX)) == NULL)) {
+        status = uct_tcp_ep_create_socket_and_connect(ep);
+        if (status != UCS_OK) {
+            return status;
+        }
+        goto out;
+    }
```

Comment:
seems like this should be a separate function

Reply:
it is not so easy, need two return values

Reply:
then need to simplify this code somehow, especially the part which initializes peer_ep inside if condition and then use it later in line 546 onward
also, i see things like uct_tcp_ep_is_self are also checked right before calling uct_tcp_ep_connect(), why need to check them again inside the function?

Reply:
simplified

> why need to check them again inside the function?

EP with connect to self must always create FD and connect to a peer's EP